File: src/main/java/com/ternbusty/takoyaki/ipc/SyncChannel.java
package com.ternbusty.takoyaki.ipc;

import com.ternbusty.takoyaki.syscall.Libc;
import com.ternbusty.takoyaki.syscall.PosixIO;

import java.lang.foreign.Arena;

public final class SyncChannel {
    public static final int MSG_INIT_READY = 0x50;
    public static final int MSG_USERMAP_PLS = 0x40;
    public static final int MSG_USERMAP_ACK = 0x41;
    public static final int MSG_CGROUP_ACK  = 0x42;

    private SyncChannel() {}

    public static int readInt32(int fd) {
        try (Arena arena = Arena.ofConfined()) {
            byte[] b = new byte[4];
            long n = PosixIO.read(arena, fd, b);
            if (n != 4) {
                throw new RuntimeException("readInt32 got " + n + " bytes (errno=" +
                        Libc.errno() + " " + Libc.strerror(Libc.errno()) + ")");
            }
            return (b[0] & 0xff) | ((b[1] & 0xff) << 8) | ((b[2] & 0xff) << 16) | ((b[3] & 0xff) << 24);
        }
    }

    public static void writeInt32(int fd, int value) {
        try (Arena arena = Arena.ofConfined()) {
            byte[] b = new byte[]{
                    (byte) (value & 0xff),
                    (byte) ((value >> 8) & 0xff),
                    (byte) ((value >> 16) & 0xff),
                    (byte) ((value >> 24) & 0xff),
            };
            long n = PosixIO.write(arena, fd, b);
            if (n != 4) throw new RuntimeException("writeInt32 wrote " + n + " bytes");
        }
    }

    /** Write a single sync byte. Used for lightweight ready signals. */
    public static void writeByte(int fd, byte value) {
        try (Arena arena = Arena.ofConfined()) {
            long n = PosixIO.write(arena, fd, new byte[]{value});
            if (n != 1) throw new RuntimeException("writeByte wrote " + n + " bytes");
        }
    }

    /** Read a single sync byte. Returns -1 on EOF. */
    public static int readByte(int fd) {
        try (Arena arena = Arena.ofConfined()) {
            byte[] b = new byte[1];
            long n = PosixIO.read(arena, fd, b);
            if (n <= 0) return -1;
            return b[0] & 0xff;
        }
    }
}

File: src/main/java/com/ternbusty/takoyaki/ipc/SyncChannel.kt
package com.ternbusty.takoyaki.ipc

import com.ternbusty.takoyaki.syscall.Libc
import com.ternbusty.takoyaki.syscall.PosixIO
import java.lang.foreign.Arena

object SyncChannel {
    const val MSG_INIT_READY = 0x50
    const val MSG_USERMAP_PLS = 0x40
    const val MSG_USERMAP_ACK = 0x41
    const val MSG_CGROUP_ACK = 0x42

    fun readInt32(fd: Int): Int {
        Arena.ofConfined().use { arena ->
            val b = ByteArray(4)
            val n = PosixIO.read(arena, fd, b)
            if (n != 4L) {
                throw RuntimeException(
                    "readInt32 got $n bytes (errno=${Libc.errno()} ${Libc.strerror(Libc.errno())})"
                )
            }
            return (b[0].toInt() and 0xff) or
                ((b[1].toInt() and 0xff) shl 8) or
                ((b[2].toInt() and 0xff) shl 16) or
                ((b[3].toInt() and 0xff) shl 24)
        }
    }

    fun writeInt32(fd: Int, value: Int) {
        Arena.ofConfined().use { arena ->
            val b = byteArrayOf(
                (value and 0xff).toByte(),
                ((value shr 8) and 0xff).toByte(),
                ((value shr 16) and 0xff).toByte(),
                ((value shr 24) and 0xff).toByte(),
            )
            val n = PosixIO.write(arena, fd, b)
            if (n != 4L) throw RuntimeException("writeInt32 wrote $n bytes")
        }
    }

    /** Write a single sync byte. Used for lightweight ready signals. */
    fun writeByte(fd: Int, value: Byte) {
        Arena.ofConfined().use { arena ->
            val n = PosixIO.write(arena, fd, byteArrayOf(value))
            if (n != 1L) throw RuntimeException("writeByte wrote $n bytes")
        }
    }

    /** Read a single sync byte. Returns -1 on EOF. */
    fun readByte(fd: Int): Int {
        Arena.ofConfined().use { arena ->
            val b = ByteArray(1)
            val n = PosixIO.read(arena, fd, b)
            if (n <= 0) return -1
            return b[0].toInt() and 0xff
        }
    }
}
